### PR TITLE
Fund password

### PR DESCRIPTION
**Pull request by nop33**
This PR addresses the following new feature:
- #560 

It also fixes the following bugs:
- #564 
- #565

and cleans up:
- #563

### CODE REVIEW

File: apps/mobile-wallet/src/components/layout/ScrollScreen.tsx [hunk 1]
```diff
@@ -69,8 +68,6 @@ const ScrollScreen = ({
   const scrollEndHandler = useAutoScrollOnDragEnd(viewRef)
   const insets = useSafeAreaInsets()
 
-  useScrollToTopOnBlur(viewRef)
-
```
Comment (nop33) [inline apps/mobile-wallet/src/components/layout/ScrollScreen.tsx:73]: I removed this to fix the annoying jump in the Settings screen happening every time we navigate to a sub-screen. Any objections?

File: apps/mobile-wallet/src/contexts/SendContext.tsx [hunk 2]
```diff
@@ -73,6 +74,7 @@ const SendContext = createContext(initialValues)
 
 export const SendContextProvider = ({ children }: { children: ReactNode }) => {
   const { triggerBiometricsAuthGuard } = useBiometricsAuthGuard()
+  const { triggerFundingPasswordAuthGuard, fundingPasswordModal } = useFundingPassword()
```
Comment (nop33) [inline apps/mobile-wallet/src/contexts/SendContext.tsx:77]: After many attempts, I was finally satisfied with an API that is similar to the biometrics one (aka `trigger...AuthGuard`. See below for implementation. This is the DRYest solution I can come up with.

File: apps/mobile-wallet/src/contexts/walletConnect/WalletConnectSessionProposalModal.tsx [hunk 3]
```diff
@@ -112,6 +111,8 @@ const WalletConnectSessionProposalModal = ({
       console.error('WC: Could not save new address', e)
 
       sendAnalytics('Error', { message: 'WC: Could not save new address' })
+    } finally {
+      keyring.clear()
```
Comment (nop33) [inline apps/mobile-wallet/src/contexts/walletConnect/WalletConnectSessionProposalModal.tsx:115]: Ensuring that secrets are cleared even when there is an exception.

File: apps/mobile-wallet/src/features/funding-password/FundingPasswordModal.tsx
Comment (nop33) [inline apps/mobile-wallet/src/features/funding-password/FundingPasswordModal.tsx:1]: I was quite impressed by the code organization at Uniswap that I was inspired to try out this `features/` folder structure. So far I like it.

File: apps/mobile-wallet/src/navigation/RootStackNavigation.tsx [hunk 4]
```diff
@@ -38,15 +38,20 @@ import ReceiveNavigation from '~/navigation/ReceiveNavigation'
...
+} from '~/persistent-storage/wallet'
 import AddressDiscoveryScreen from '~/screens/AddressDiscoveryScreen'
 import EditAddressScreen from '~/screens/Addresses/Address/EditAddressScreen'
 import NewAddressScreen from '~/screens/Addresses/Address/NewAddressScreen'
 import ContactScreen from '~/screens/Addresses/Contact/ContactScreen'
 import EditContactScreen from '~/screens/Addresses/Contact/EditContactScreen'
 import NewContactScreen from '~/screens/Addresses/Contact/NewContactScreen'
 import CustomNetworkScreen from '~/screens/CustomNetworkScreen'
-import LandingScreen from '~/screens/LandingScreen'
+import LandingScreen, { CoolAlephiumCanvas } from '~/screens/LandingScreen'
```
Comment (nop33) [inline apps/mobile-wallet/src/navigation/RootStackNavigation.tsx:54]: You'l probably disagree with the name but whatever :P

File: apps/mobile-wallet/src/navigation/RootStackNavigation.tsx [hunk 5]
```diff
@@ -94,7 +98,6 @@ const RootStackNavigation = () => {
     <GestureHandlerRootView style={{ flex: 1 }}>
       <Host>
         <NavigationContainer ref={rootStackNavigationRef} onStateChange={handleStateChange} theme={themeNavigator}>
-          <AppUnlockHandler />
```
Comment (nop33) [inline apps/mobile-wallet/src/navigation/RootStackNavigation.tsx:97]: I moved this to the bottom so that it is rendered above everything else

File: apps/mobile-wallet/src/navigation/RootStackNavigation.tsx [hunk 6]
```diff
@@ -147,22 +152,30 @@ const AppUnlockHandler = () => {
...
   const [needsWalletUnlock, setNeedsWalletUnlock] = useState(false)
+  const [isAuthModalVisible, setIsAuthModalVisible] = useState(false)
 
-  const initializeWallet = useCallback(
-    (wallet: WalletMetadata) => {
-      dispatch(walletUnlocked(wallet))
+  const { width, height } = Dimensions.get('window')
+  const [dimensions, setDimensions] = useState({ width, height })
 
-      // TODO: Remove when resetting navigation to LandingScreen is removed
```
Comment (nop33) [inline apps/mobile-wallet/src/navigation/RootStackNavigation.tsx:160]: Done :wink:

File: apps/mobile-wallet/src/navigation/RootStackNavigation.tsx [hunk 7]
```diff
@@ -208,15 +214,16 @@ const AppUnlockHandler = () => {
...
     }
-  }, [initializeWallet, isWalletUnlocked, lastNavigationState, navigation, triggerBiometricsAuthGuard])
+  }, [initializeAppWithStoredWallet, isWalletUnlocked, navigation, triggerBiometricsAuthGuard])
 
   useEffect(() => {
+    if (!settingsLoadedFromStorage) return
+
     const handleAppStateChange = (nextAppState: AppStateStatus) => {
       if (nextAppState === 'background' && isWalletUnlocked && !isCameraOpen) {
-        // TODO: Show auth modal without first resetting navigation
```
Comment (nop33) [inline apps/mobile-wallet/src/navigation/RootStackNavigation.tsx:216]: Also done!

File: apps/mobile-wallet/src/persistent-storage/wallet.ts [hunk 8]
```diff
@@ -58,25 +59,30 @@ export const generateAndStoreWallet = async (
...
+      isMnemonicBackedUp,
+      firstAddress: {
+        index: 0,
+        hash: firstAddressHash
+      }
     }
+  } catch (e) {
+    throw new Error(getHumanReadableError(e, ''))
+  } finally {
+    keyring.clear()
```
Comment (nop33) [inline apps/mobile-wallet/src/persistent-storage/wallet.ts:85]: Same thing as before, making sure secrets are cleared no matter what

File: apps/mobile-wallet/src/screens/BackupMnemonic/VerifyMnemonicScreen.tsx
Comment (nop33) [inline apps/mobile-wallet/src/screens/BackupMnemonic/VerifyMnemonicScreen.tsx:1]: idk if this bug exists in prod or not, but verification of words is skipped without this fix.

File: apps/mobile-wallet/src/screens/new-wallet/DecryptScannedMnemonicScreen.tsx [hunk 9]
```diff
@@ -100,7 +101,12 @@ const DecryptScannedMnemonicScreen = ({ navigation }: DecryptScannedMnemonicScre
           sendAnalytics('Error', { message })
         }
 
-        resetNavigation(navigation, deviceHasEnrolledBiometrics ? 'AddBiometricsScreen' : 'NewWalletSuccessScreen')
+        resetNavigation(
+          navigation,
+          deviceHasEnrolledBiometrics && !biometricsRequiredForAppAccess
+            ? 'AddBiometricsScreen'
+            : 'NewWalletSuccessScreen'
+        )
```
Comment (nop33) [inline apps/mobile-wallet/src/screens/new-wallet/DecryptScannedMnemonicScreen.tsx:109]: Only navigate to the biometrics screen if biometrics is NOT enabled

File: apps/mobile-wallet/src/store/settingsSlice.ts
Comment (nop33) [inline apps/mobile-wallet/src/store/settingsSlice.ts:1]: I should have probably created a slice in the features directory but at the end I was too lazy

File: apps/mobile-wallet/src/navigation/RootStackNavigation.tsx (hunk 4 above)
Comment (mvaivre) [reply apps/mobile-wallet/src/navigation/RootStackNavigation.tsx:54]: Mouhaha, I like it ;)

**Review by mvaivre — COMMENTED** [review]
I've started to review and try to test this on iOS, but for now I'm stuck at the landing screen :/ I'm gonna check the native logs.

One remark: we were talking about calling this password "Fund password", not "Funding password". 
I've researched a bit around, and "Fund password" is used in most of the main exchanges.

File: apps/mobile-wallet/src/features/funding-password/FundingPasswordModal.tsx
Comment (mvaivre) [reply apps/mobile-wallet/src/features/funding-password/FundingPasswordModal.tsx:1]: That's a whole paradigm shift. I've considered this before and was not convinced in the end because sometimes (often?) the limits between 2 features aren't clear.
 
We could try and see how we could reorganise our files per features, see how we would do it.

Comment (nop33) [reply apps/mobile-wallet/src/features/funding-password/FundingPasswordModal.tsx:1]: Agreed 🤝

**Comment by nop33** [comment]
> I've started to review and try to test this on iOS, but for now I'm stuck at the landing screen :/ I'm gonna check the native logs.

Strange. It works on the simulator:


https://github.com/alephium/alephium-frontend/assets/1579899/b54b8644-5291-4e61-b840-2b2ef66c7a24

File: apps/mobile-wallet/src/components/layout/ScrollScreen.tsx (hunk 1 above)
Comment (mvaivre) [reply apps/mobile-wallet/src/components/layout/ScrollScreen.tsx:73]: Sounds good to me.

**Comment by mvaivre** [comment]
I'm not stuck on the landing screen anymore! Nice work mate!! 🤙 👏 

Some suggestions:
- I would activate use of biometrics to send transaction by default
- I would highlight the message that says that biometrics can be enabled 

image

I would also show a full size bottom modal that pops up for existing users on app launch, which explains that the PIN has been replaced by a Fund Password, and suggest them to set such a password to improve security. We can mention that the new settings can be found in the settings screen. Two choices: "Later", "Set Fund Password". 
We could show this modal only once... or twice? In case the use accidentally swiped down? WDYT?

About the explanations of what this password is, I would rephrase and include other critical actions ("additional authentication layer for critical operations involving the safety of your funds"). This would include showing the mnemonic for instance.
image

**Comment by nop33** [comment]
> One remark: we were talking about calling this password "Fund(s) password", not "Funding password".

Done ✅ 

> I would activate use of biometrics to send transaction by default

Hm, I wouldn't. In the wallet creation flow we ask the user if they want to enable biometrics for app access or not. If they say "no", we show them the `BiometricsWarningModal` bottom modal with a message that reads _"If you don't turn on biometrics, anyone who gains access to your device can open the app and steal your funds."_. If we'd enable the `biometricsRequiredForTransactions` without the consent of the user, then the above doesn't make sense.

An option would be to enable this setting when the user activates the biometrics in the wallet creation flow, as we currently do for the `biometricsRequiredForAppAccess` setting. So basically enabling both settings at once.

> I would highlight the message that says that biometrics can be enabled

I'll work on it 👍 

> I would also show a full size bottom modal that pops up for existing users on app launch, which explains that the PIN has been replaced by a Fund Password, and suggest them to set such a password to improve security. We can mention that the new settings can be found in the settings screen. Two choices: "Later", "Set Fund Password".

That's a good suggestion! I wouldn't show a full size one though, I'd size it according to its contents.

> We could show this modal only once... or twice? In case the use accidentally swiped down? WDYT?

I'd say once is enough.

> About the explanations of what this password is, I would rephrase and include other critical actions ("additional authentication layer for critical operations involving the safety of your funds"). This would include showing the mnemonic for instance.

I personally find this messaging a bit too general. I kinda prefer the one I recommended. It's more "straight to the point". Showing the mnemonic is the same as your funds leaving your wallet anyway!

**Comment by mvaivre** [comment]
> So basically enabling both settings at once.
That's what I meant :) Basically, *if biometrics are enabled*, then we ask for biometric auth when sending a TX by default.

> I personally find this messaging a bit too general. I kinda prefer the one I recommended. It's more "straight to the point". 
I've quickly asked around to a couple of friends, and they both like that the words *critical* and *safety* make it clearer that it's for all important actions, and give a better feel that the user is strongly encouraged to use it. I'm asking Maud / Vlad for an additional quick opinion :)

**Comment by MaudSimon** [comment]
Hey my personal thoughts on the various messages: 

> "If you don't turn on biometrics, anyone who gains access to your device can open the app and steal your funds.".
I'd replace by _"If you don't turn on biometrics, anyone who gains access to your device can open the app and access your funds."_

I would also go for this "Additional authentication layer for critical operations involving the safety of your funds." if you wanted to be more clear: "Additional authentication layer for critical operations involving the safety of your funds such as revealing your seed phrase or sending funds."

**Comment by VDAODAO** [comment]
As it comes in the settings where the taxonomy is the following: 

"
**Feature Name Bold**
3-Words description
"

I would suggest to go with the following: 

**Set Fund Password**
Additional authentication layer

And I'd suggest to add the full details on why/what is entailed by that at the point of onboarding, where you can put all the places this is going to be used as bulletpoints/list: 

- additional security before moving funds
- the thing with the seed
- etc....

**Comment by mvaivre** [comment]
Thanks both of you for your feedback!

I'd suggest:
- To use the short version for the settings ("Additional authentication layer")
- To use a more complete version in the bottom modal and the onboarding flow ("Additional authentication layer for critical operations involving the safety of your funds such as revealing your seed phrase or sending funds.)

**Comment by mvaivre** [comment]
Tested on the simulator:

- Checked out master
- Deleted the wallet
- Imported a wallet
- Checked out you branch
- Everything seems to work! 🥳 

However, on my device I'm getting stuck again for a long time (a few mis) on the landing screen, before the app actually starts. But this happens on master as well 🤔

**Comment by nop33** [comment]
> Case 1:

That's normal. You first need to back up your mnemonic. Only then you set the password. That's the flow we discussed, right? You always first get notified for backing up the mnemonic. That has priority.

> Nit: It's asking for the Fund Password but I see a weird re-rendering artefact
> Nit: I enter the password, the words are showing, but I see the same re-rendering issue

I'll check this

> Side comment: on my device I'm getting stuck again for a long time (a few minutes) on the landing screen, before the app actually starts. But this happens on master as well 🤔 Have you ever encountered this on Android?

I have not experienced that, no. I'll try on the simulator on iOS

**Comment by mvaivre** [comment]
> That's the flow we discussed, right?

**Comment by nop33** [comment]
I've also fixed being stuck in splash screen.

As for the small artefacts, I'd like to focus on this another time 🙏

File: apps/mobile-wallet/src/features/fund-password/FundPasswordScreen.tsx [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+import { resetNavigation } from '~/utils/navigation'
+
+interface FundPasswordScreenProps
+  extends StackScreenProps<RootStackParamList, 'FundPasswordScreen'>,
+    ScrollScreenProps {}
+
+const FundPasswordScreen = ({ navigation, ...props }: FundPasswordScreenProps) => {
+  const cameFromBackupScreen = props.route.params.origin === 'backup'
+  const { setHeaderOptions } = useHeaderContext()
+  const { data: currentFundPassword } = useAsyncData(getFundPassword)
```
Comment (mvaivre) [inline apps/mobile-wallet/src/features/fund-password/FundPasswordScreen.tsx:48]: Could use custom hook here (cf. previous comment)

File: apps/mobile-wallet/src/features/fund-password/fundPasswordStorage.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
+/*
```
Comment (mvaivre) [inline apps/mobile-wallet/src/features/fund-password/fundPasswordStorage.ts:1]: Q: You still like the `features/***` way of ordering files? Do you see any clear pros and cons so far?

**Review by mvaivre — COMMENTED** [review]
Part 1 👏

File: apps/mobile-wallet/src/features/fund-password/fundPasswordStorage.ts (hunk 11 above)
Comment (nop33) [reply apps/mobile-wallet/src/features/fund-password/fundPasswordStorage.ts:1]: I like it yes. Everything regarding the feature that is not being used in any other part of the app is contained within one folder.

File: apps/mobile-wallet/src/features/fund-password/FundPasswordReminderModal.tsx [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  const handleSetPasswordPress = () => {
+    onClose()
+
+    navigation.navigate('FundPasswordScreen', { origin: 'settings' })
+  }
+
+  return (
+    <Portal>
```
[the remark this answers is not in the document]
Comment (nop33) [reply apps/mobile-wallet/src/features/fund-password/FundPasswordReminderModal.tsx:64]: It's a good point, it probably should. It will require quite some cleanup in our code base and some good testing. Would you mind leaving this in a separate PR? I'll make an issue for it.

[the remark this answers is not in the document]
Comment (mvaivre) [reply apps/mobile-wallet/src/features/fund-password/FundPasswordReminderModal.tsx:64]: Sounds good!

[the remark this answers is not in the document]
Comment (nop33) [reply apps/mobile-wallet/src/features/fund-password/FundPasswordReminderModal.tsx:64]: https://github.com/alephium/alephium-frontend/issues/598

File: apps/mobile-wallet/src/screens/BackupMnemonic/VerifyMnemonicScreen.tsx [hunk 13]
```diff
@@ -74,17 +76,24 @@ const VerifyMnemonicScreen = ({ navigation, ...props }: VerifyMnemonicScreenProp
...
+        await updateStoredWalletMetadata({ isMnemonicBackedUp: true })
+        dispatch(mnemonicBackedUp())
 
-      sendAnalytics('Backed-up mnemonic')
+        sendAnalytics({ event: 'Backed-up mnemonic' })
+      } catch (error) {
+        const message = 'Could not confirm backup'
+
+        showExceptionToast(error, message)
+        sendAnalytics({ type: 'error', error, message })
```
Comment (mvaivre) [inline apps/mobile-wallet/src/screens/BackupMnemonic/VerifyMnemonicScreen.tsx:88]: Double checking: no risk of leaking the mnemonic here?

File: apps/mobile-wallet/src/screens/new-wallet/AddBiometricsScreen.tsx [hunk 14]
```diff
@@ -53,16 +53,16 @@ const AddBiometricsScreen = ({ navigation, ...props }: AddBiometricsScreenProps)
   const skipAddressDiscovery = method === 'create' || addressIds.length > 1
 
   const activateBiometrics = () => {
-    dispatch(biometricsToggled())
+    dispatch(allBiometricsEnabled())
```
Comment (mvaivre) [inline apps/mobile-wallet/src/screens/new-wallet/AddBiometricsScreen.tsx:56]: Nit: I find `allBiometricsEnabled` a bit confusing. Is there "multiple biometrics"? Couldn't we just say `biometricsEnabled`? :)

**Review by mvaivre — COMMENTED** [review]
Very nice job!! 🤙

File: apps/mobile-wallet/src/screens/BackupMnemonic/VerifyMnemonicScreen.tsx (hunk 13 above)
Comment (nop33) [reply apps/mobile-wallet/src/screens/BackupMnemonic/VerifyMnemonicScreen.tsx:88]: No risk

File: apps/mobile-wallet/src/screens/new-wallet/AddBiometricsScreen.tsx (hunk 14 above)
Comment (nop33) [reply apps/mobile-wallet/src/screens/new-wallet/AddBiometricsScreen.tsx:56]: There is:
- `biometricsRequiredForAppAccess`
- `biometricsRequiredForTransactions`

File: apps/mobile-wallet/src/persistent-storage/utils.ts [hunk 15]
```diff
@@ -16,38 +16,86 @@ You should have received a copy of the GNU Lesser General Public License
...
-    throw new Error(error)
+    await SecureStore.deleteItemAsync(key, defaultSecureStoreConfig)
+  } catch (error) {
+    sendAnalytics({
+      type: 'error',
+      message: showDefaultErrorMessage ? `Could not delete ${key} from secure storage` : errorMessage
+    })
+    throw error
   }
 }
```
[the remark this answers is not in the document]
Comment (nop33) [reply apps/mobile-wallet/src/persistent-storage/utils.ts:101]: This was done by design, I wanted the reportable error message to be an explicit thing. I don't want it to have a default because some storage keys might contain sensitive info, like the key of the private key entry (which contains the address hash). That's why I made the error message a mandatory arg, but also introduced the boolean arg `showDefaultErrorMessage` so that the developer will first check the implementation of the function before calling it. This was done to avoid leaking any unwanted sensitive info by accident. I favored privacy/security over code prettiness here.

File: apps/mobile-wallet/src/hooks/usePassword.ts [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { useState } from 'react'
+
```
[the remark this answers is not in the document]
Comment (nop33) [reply apps/mobile-wallet/src/hooks/usePassword.ts:20]: Fixed

File: apps/mobile-wallet/src/features/fund-password/useFundPasswordGuard.tsx [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+const useFundPasswordGuard = () => {
+  const usesFundPassword = useAppSelector((s) => s.settings.usesFundPassword)
+
+  const [onCorrectPasswordCallback, setOnCorrectPasswordCallback] = useState<() => void>(() => () => null)
+  const [isFundPasswordModalOpen, setIsFundPasswordModalOpen] = useState(false)
+
+  const triggerFundPasswordAuthGuard = useCallback(
+    ({ successCallback }: Pick<FundPasswordModalProps, 'successCallback'>) => {
+      if (usesFundPassword) {
+        setOnCorrectPasswordCallback(() => () => successCallback())
```
[the remark this answers is not in the document]
Comment (nop33) [reply apps/mobile-wallet/src/features/fund-password/useFundPasswordGuard.tsx]: Fixed

File: apps/mobile-wallet/src/features/fund-password/FundPasswordModal.tsx [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+import useFundPassword from '~/features/fund-password/useFundPassword'
+import { useAppSelector } from '~/hooks/redux'
+import usePassword from '~/hooks/usePassword'
+
+export interface FundPasswordModalProps extends Pick<BottomModalProps, 'isOpen' | 'onClose'> {
+  successCallback: () => void
+}
+
+const FundPasswordModal = ({ successCallback, ...props }: FundPasswordModalProps) => {
+  const usesFundPassword = useAppSelector((s) => s.settings.usesFundPassword)
```
[the remark this answers is not in the document]
Comment (nop33) [reply apps/mobile-wallet/src/features/fund-password/FundPasswordModal.tsx]: Fixed

File: apps/mobile-wallet/src/persistent-storage/utils.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (mvaivre) [reply apps/mobile-wallet/src/persistent-storage/utils.ts:101]: Makes sense. Thanks for the explanations 🙏

**Comment by mvaivre** [comment]
**Testing session ⚙️ **

I'm not a super fan of the current flow to edit or deactivate the fund password. It's weird to "pseudo-deactivate" the Fund Password (tapping on the switch), to then access a page which let us actually deactivate it.
If a password is set, the settings menu should redirect to a page to either edit or deactivate the fund password.
Tapping on one of the 2 options should prompt the "fund password" modal. I'm gonna work on this in a new branch.

Also, to make brute forcing less trivial, the password check should be done when tapping on submit, not before:
image

_Continuing..._

**Review by mvaivre — APPROVED** [review]
OHYEAH
